stdlib/collections/__init__.pyi

```diff
@@ -391,6 +391,9 @@ class defaultdict(dict[_KT, _VT], Generic[_KT, _VT]):
     def __missing__(self, __key: _KT) -> _VT: ...
     def __copy__(self) -> Self: ...
     def copy(self) -> Self: ...
+    if sys.version_info >= (3, 9):
+        def __or__(self, other: Mapping[_T1, _T2]) -> defaultdict[_KT | _T1, _VT | _T2]: ...
+        def __ror__(self, other: Mapping[_T1, _T2]) -> defaultdict[_KT | _T1, _VT | _T2]: ...
```

Comment:
If you have a subclass of `defaultdict`, then `__or__` will actually return an instance of the subclass (note this _differs_ from how `dict.__or__` behaves):

```pycon
>>> from collections import defaultdict
>>> class Foo(defaultdict): ...
...
>>> type(Foo(int) | defaultdict(int))
<class '__main__.Foo'>
```

We can't express this fully without higher-kinded typevars (https://github.com/python/typing/issues/548), but we _can_ give a more specific return type in the specific case where a mapping that has the same type parameters is being `__or__`'d.

Also: the second parameter should be named `value`, not `other`:

```pycon
>>> from collections import defaultdict
>>> help(defaultdict.__or__)
Help on wrapper_descriptor:

__or__(self, value, /)
    Return self|value.
```

And it should be positional-only:

```pycon
>>> from collections import defaultdict
>>> x = defaultdict(int)
>>> x.__or__(value={})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: wrapper __or__() takes no keyword arguments
```

Putting all this together, I think the signature should be:

```suggestion
        @overload
        def __or__(self, __value: Mapping[_KT, _VT]) -> Self: ...
        @overload
        def __or__(self, __value: Mapping[_T1, _T2]) -> defaultdict[_KT | _T1, _VT | _T2]: ...
        @overload
        def __or__(self, __value: Mapping[_KT, _VT]) -> Self: ..
        @overload
        def __ror__(self, __value: Mapping[_T1, _T2]) -> defaultdict[_KT | _T1, _VT | _T2]: ...
```

Reply:
Probably the type of the other parameter should be `_typeshed.SupportsKeysAndGetItem`, not `Mapping`, right?

Reply:
> Probably the type of the other parameter should be `_typeshed.SupportsKeysAndGetItem`, not `Mapping`, right?

Doesn't look like it?

```pycon
>>> from collections import defaultdict
>>> class Foo:
...     def __init__(self):
...         self.data = {}
...     def keys(self):
...         return self.data.keys()
...     def __getitem__(self, key):
...         return self.data[key]
...
>>> defaultdict(int) | Foo()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for |: 'collections.defaultdict' and 'Foo'
```

Reply:
@JelleZijlstra @AlexWaygood defaultdict actually only supports subclasses of dict: https://github.com/python/cpython/blob/388b5daa523b828dc0f7e2a1a6886bebc20833ba/Modules/_collectionsmodule.c#L2136
However, we cannot declare it here as `dict`, because `dict` is declared to accept `Mapping` in typeshed, so this causes a violation of Liskov substitution. This also seems wrong. The code only accept dicts: https://github.com/python/cpython/blob/388b5daa523b828dc0f7e2a1a6886bebc20833ba/Objects/dictobject.c#L3606
Quick demo:
```python
>>> from collections.abc import Mapping
>>>> class D(Mapping):
    def __getitem__(self, key):
        raise KeyError
    def __len__(self):
        return 0
    def __iter__(self):
        return iter(())

>>> {"a": 1} | D()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for |: 'dict' and 'D'
```

Reply:
> @JelleZijlstra @AlexWaygood defaultdict actually only supports subclasses of dict:

Yes we had just come to the same conclusion :)

> However, we cannot declare it here as `dict`, because `dict` is declared to accept `Mapping` in typeshed, so this causes a violation of Liskov substitution. This also seems wrong. The code only accept dicts: https://github.com/python/cpython/blob/388b5daa523b828dc0f7e2a1a6886bebc20833ba/Objects/dictobject.c#L3606

Good catch! Want to update `dict.__(r)or__` in this PR as well?